File: src/main/ele_action/handlers/file_render/adapter/SVGAdapter.js
// SVGAdapter.js
import { IAdapter } from './IAdapter';
import { OverviewStorage, ImageStorage } from '../utils';

const displayScale = 10;

export class SVGAdapter extends IAdapter {
  constructor(config, quality = 'high', useAppLinks = false) {
    super();
    this.config = config;
    this.useAppLinks = useAppLinks;
    this.imageQuality = quality;

    this.pages = [];
    this.currentPageIndex = -1;
    this.currentPage = null;
    this.transformStack = [];
    this.currentTransform = { a: 1, b: 0, c: 0, d: 1, e: 0, f: 0 };

    const [width, height] = this.parsePageSize(config.pageSize);

    if (config.landscape) {
      this.pageWidth = height;
      this.pageHeight = width;
    } else {
      this.pageWidth = width;
      this.pageHeight = height;
    }

    console.log('SVG Page Size:', {
      width: this.pageWidth,
      height: this.pageHeight,
      landscape: config.landscape
    });

    this.createNewPage();
  }

  parsePageSize(pageSize) {
    const parts = pageSize.split(':');
    if (parts.length === 2) {
      const [width, height] = parts[1].split(',').map(Number);
      return [width, height];
    }
    return [595, 842];
  }

  createNewPage() {
    this.currentPageIndex++;
    this.currentPage = {
      index: this.currentPageIndex,
      elements: []
    };
    this.pages.push(this.currentPage);
  }

  addPage() {
    this.createNewPage();
  }

  getPageSize() {
    return {
      width: this.pageWidth,
      height: this.pageHeight
    };
  }

  setTransform({ a = 1, b = 0, c = 0, d = 1, e = 0, f = 0 }) {
    this.currentTransform = { a, b, c, d, e, f };
  }

  saveState() {
    this.transformStack.push({ ...this.currentTransform });
  }

  restoreState() {
    if (this.transformStack.length > 0) {
      this.currentTransform = this.transformStack.pop();
    }
  }

  applyTransform(x, y) {
    const { a, b, c, d, e, f } = this.currentTransform;
    return {
      x: a * x + c * y + e,
      y: b * x + d * y + f
    };
  }

  setLineStyle({ width, color }) {
    this.currentLineWidth = width;
    this.currentLineColor = color;
  }

  drawImage({ data, x, y, width, height, rotation = 0 }) {
    const imagePathKey = data.path.replaceAll('\\', '');

    let imageSource;
    if (this.useAppLinks) {
      const quality = this.imageQuality === 'high' ? 'high' : 'low';
      imageSource = `cardrac://image/${imagePathKey}?quality=${quality}`;
    } else {
      const storage = this.imageQuality === 'low' ? OverviewStorage : ImageStorage;
      imageSource = storage[imagePathKey] || '';
    }

    let adjustedX = x;
    let adjustedY = y;

    if (rotation === 180) {
      adjustedX = x - width;
      adjustedY = y + height;
    }
    const transformed = this.applyTransform(adjustedX, adjustedY);
    this.currentPage.elements.push({
      type: 'image',
      href: imageSource,
      x: transformed.x,
      y: transformed.y,
      width,
      height,
      rotation,
      path: data.path
    });
  }

  drawText({ text, x, y, size = 12 }) {
    const transformed = this.applyTransform(x, y);
    this.currentPage.elements.push({
      type: 'text',
      text,
      x: transformed.x,
      y: transformed.y,
      size: Math.floor(size / 2.5)
    });
  }

  drawLine({ x1, y1, x2, y2, dash }) {
    const transformed1 = this.applyTransform(x1, y1);
    const transformed2 = this.applyTransform(x2, y2);

    this.currentPage.elements.push({
      type: 'line',
      x1: transformed1.x,
      y1: transformed1.y,
      x2: transformed2.x,
      y2: transformed2.y,
      dash,
      width: this.currentLineWidth || 1,
      color: this.currentLineColor || '#000000'
    });
  }

  fillRect({ x, y, width, height, color }) {
    const transformed = this.applyTransform(x, y);
    this.currentPage.elements.push({
      type: 'rect',
      x: transformed.x,
      y: transformed.y,
      width,
      height,
      color
    });
  }

  // ✅ 修改：按绘制顺序渲染，不分组
  generatePageSVG(page) {
    const svg = `<svg 
    width="${this.pageWidth * displayScale}" 
    height="${this.pageHeight * displayScale}" 
    viewBox="0 0 ${this.pageWidth * displayScale} ${this.pageHeight * displayScale}"
    xmlns="http://www.w3.org/2000/svg">
    <rect width="${this.pageWidth * displayScale}" height="${this.pageHeight * displayScale}" fill="white"/>
    ${page.elements.map(el => this.renderElement(el)).join('\n')}
  </svg>`;
    return svg;
  }

  renderElement(el) {
    const s = displayScale;

    switch (el.type) {
      case 'image':
        if (!el.href) return '';
        const centerX = (el.x + el.width / 2) * s;
        const centerY = (el.y + el.height / 2) * s;

        return `<image 
    href="${el.href}" 
    x="${el.x * s}" 
    y="${el.y * s}" 
    width="${el.width * s}" 
    height="${el.height * s}" 
    preserveAspectRatio="none"
    transform="rotate(${el.rotation}, ${centerX}, ${centerY})" />`;

      case 'text':
        const escapedText = el.text
          .replace(/&/g, '&amp;')
          .replace(/</g, '<')
          .replace(/>/g, '>')
          .replace(/"/g, '&quot;')
          .replace(/'/g, '&apos;');
        return `<text x="${el.x * s}" y="${el.y * s}" font-size="${el.size * s}" font-family="Arial">${escapedText}</text>`;

      case 'line':
        return `<line 
          x1="${el.x1 * s}" 
          y1="${el.y1 * s}" 
          x2="${el.x2 * s}" 
          y2="${el.y2 * s}" 
          stroke="${el.color || '#000000'}"
          stroke-width="${el.width * s}"
          ${el.dash ? `stroke-dasharray="${el.dash.map(d => d * s).join(',')}"` : ''} />`;

      case 'rect':
        return `<rect 
          x="${el.x * s}" 
          y="${el.y * s}" 
          width="${el.width * s}" 
          height="${el.height * s}" 
          fill="rgb(${el.color.r},${el.color.g},${el.color.b})" />`;

      default:
        return '';
    }
  }

  finalize() {
    const validPages = this.pages.filter(page => page.elements.length > 0);

    if (validPages.length === 0) {
      throw new Error('No pages to export');
    }

    if (validPages.length === 1) {
      return this.generatePageSVG(validPages[0]);
    } else {
      return validPages.map(page => this.generatePageSVG(page));
    }
  }
}

File: server/file_render/adapter/svg_adapter.go
package adapter

import (
    "encoding/base64"
    "fmt"
    "strings"
)

const displayScale = 10

// SVGAdapter SVG 适配器
type SVGAdapter struct {
    config           map[string]interface{}
    useAppLinks      bool
    imageQuality     string
    pages            []SVGPage
    currentPageIndex int
    currentPage      *SVGPage
    transformStack   []Transform
    currentTransform Transform
    pageWidth        float64
    pageHeight       float64
    currentLineWidth float64
    currentLineColor string
}

// SVGPage SVG 页面
type SVGPage struct {
    Index    int          `json:"index"`
    Elements []SVGElement `json:"elements"`
}

// SVGElement SVG 元素
type SVGElement struct {
    Type     string                 `json:"type"`
    Data     map[string]interface{} `json:"data"`
    X        float64                `json:"x,omitempty"`
    Y        float64                `json:"y,omitempty"`
    Width    float64                `json:"width,omitempty"`
    Height   float64                `json:"height,omitempty"`
    Rotation float64                `json:"rotation,omitempty"`
}

// NewSVGAdapter 创建 SVG 适配器
func NewSVGAdapter(config map[string]interface{}, quality string, useAppLinks bool) *SVGAdapter {
    landscape, _ := config["landscape"].(bool)
    pageWidth, _ := config["pageWidth"].(float64)
    pageHeight, _ := config["pageHeight"].(float64)

    width := pageWidth
    height := pageHeight

    if landscape {
        width, height = height, width
    }

    adapter := &SVGAdapter{
        config:       config,
        useAppLinks:  useAppLinks,
        imageQuality: quality,
        pages:        []SVGPage{},
        currentPageIndex: -1,
        transformStack: []Transform{},
        currentTransform: Transform{A: 1, B: 0, C: 0, D: 1, E: 0, F: 0},
        pageWidth:    width,
        pageHeight:   height,
    }

    adapter.createNewPage()
    return adapter
}

// createNewPage 创建新页面
func (s *SVGAdapter) createNewPage() {
    s.currentPageIndex++
    s.currentPage = &SVGPage{
        Index:    s.currentPageIndex,
        Elements: []SVGElement{},
    }
    s.pages = append(s.pages, *s.currentPage)
}

// AddPage 添加新页面
func (s *SVGAdapter) AddPage() error {
    s.createNewPage()
    return nil
}

// GetPageSize 获取页面尺寸
func (s *SVGAdapter) GetPageSize() PageSize {
    return PageSize{
        Width:  s.pageWidth,
        Height: s.pageHeight,
    }
}

// SetTransform 设置变换
func (s *SVGAdapter) SetTransform(transform Transform) error {
    s.currentTransform = transform
    return nil
}

// SaveState 保存状态
func (s *SVGAdapter) SaveState() error {
    s.transformStack = append(s.transformStack, s.currentTransform)
    return nil
}

// RestoreState 恢复状态
func (s *SVGAdapter) RestoreState() error {
    if len(s.transformStack) > 0 {
        s.currentTransform = s.transformStack[len(s.transformStack)-1]
        s.transformStack = s.transformStack[:len(s.transformStack)-1]
    }
    return nil
}

// applyTransform 应用变换
func (s *SVGAdapter) applyTransform(x, y float64) (float64, float64) {
    t := s.currentTransform
    newX := t.A*x + t.C*y + t.E
    newY := t.B*x + t.D*y + t.F
    return newX, newY
}

// SetLineStyle 设置线条样式
func (s *SVGAdapter) SetLineStyle(style LineStyle) error {
    s.currentLineWidth = style.Width
    s.currentLineColor = style.Color
    return nil
}

// DrawImage 绘制图片
func (s *SVGAdapter) DrawImage(options ImageOptions) error {
    // 构建图片源
    imagePathKey := strings.ReplaceAll(options.Data.Path, "\\", "")
    var imageSource string

    if s.useAppLinks {
        quality := "low"
        if s.imageQuality == "high" {
            quality = "high"
        }
        imageSource = fmt.Sprintf("cardrac://image/%s?quality=%s", imagePathKey, quality)
    } else {
        // 使用 base64 数据
        imageSource = options.Data.Base64
    }

    adjustedX := options.X
    adjustedY := options.Y

    if options.Rotation == 180 {
        adjustedX = options.X - options.Width
        adjustedY = options.Y + options.Height
    }

    transformedX, transformedY := s.applyTransform(adjustedX, adjustedY)

    s.currentPage.Elements = append(s.currentPage.Elements, SVGElement{
        Type:     "image",
        X:        transformedX,
        Y:        transformedY,
        Width:    options.Width,
        Height:   options.Height,
        Rotation: options.Rotation,
        Data: map[string]interface{}{
            "href": imageSource,
            "path": options.Data.Path,
        },
    })

    return nil
}

// DrawText 绘制文本
func (s *SVGAdapter) DrawText(options TextOptions) error {
    transformedX, transformedY := s.applyTransform(options.X, options.Y)

    s.currentPage.Elements = append(s.currentPage.Elements, SVGElement{
        Type: "text",
        X:    transformedX,
        Y:    transformedY,
        Data: map[string]interface{}{
            "text": options.Text,
            "size": options.Size / 2.5, // 调整字体大小
        },
    })

    return nil
}

// DrawLine 绘制线条
func (s *SVGAdapter) DrawLine(options LineOptions) error {
    x1, y1 := s.applyTransform(options.X1, options.Y1)
    x2, y2 := s.applyTransform(options.X2, options.Y2)

    s.currentPage.Elements = append(s.currentPage.Elements, SVGElement{
        Type: "line",
        Data: map[string]interface{}{
            "x1":    x1,
            "y1":    y1,
            "x2":    x2,
            "y2":    y2,
            "dash":  options.Dash,
            "width": s.currentLineWidth,
            "color": s.currentLineColor,
        },
    })

    return nil
}

// FillRect 填充矩形
func (s *SVGAdapter) FillRect(options RectOptions) error {
    transformedX, transformedY := s.applyTransform(options.X, options.Y)

    s.currentPage.Elements = append(s.currentPage.Elements, SVGElement{
        Type:   "rect",
        X:      transformedX,
        Y:      transformedY,
        Width:  options.Width,
        Height: options.Height,
        Data: map[string]interface{}{
            "color": options.Color,
        },
    })

    return nil
}

// Finalize 完成并生成 SVG
func (s *SVGAdapter) Finalize() (interface{}, error) {
    validPages := []SVGPage{}
    for _, page := range s.pages {
        if len(page.Elements) > 0 {
            validPages = append(validPages, page)
        }
    }

    if len(validPages) == 0 {
        return s.generateEmptySVG(), nil
    }

    if len(validPages) == 1 {
        return s.generatePageSVG(&validPages[0]), nil
    }

    svgs := []string{}
    for i := range validPages {
        svgs = append(svgs, s.generatePageSVG(&validPages[i]))
    }
    return svgs, nil
}

// generatePageSVG 生成页面 SVG
func (s *SVGAdapter) generatePageSVG(page *SVGPage) string {
    width := s.pageWidth * displayScale
    height := s.pageHeight * displayScale

    var elements strings.Builder
    for _, el := range page.Elements {
        elements.WriteString(s.renderElement(el))
        elements.WriteString("\n")
    }

    svg := fmt.Sprintf(`<svg width="%.2f" height="%.2f" viewBox="0 0 %.2f %.2f" xmlns="http://www.w3.org/2000/svg">
<rect width="%.2f" height="%.2f" fill="white"/>
%s
</svg>`, width, height, width, height, width, height, elements.String())

    return svg
}

// generateEmptySVG 生成空 SVG
func (s *SVGAdapter) generateEmptySVG() string {
    width := s.pageWidth * 10
    height := s.pageHeight * 10

    return fmt.Sprintf(`<svg width="%.2f" height="%.2f" viewBox="0 0 %.2f %.2f" xmlns="http://www.w3.org/2000/svg">
<rect width="%.2f" height="%.2f" fill="white"/>
<text x="%.2f" y="%.2f" text-anchor="middle" font-size="48" fill="#999">No content to display</text>
</svg>`, width, height, width, height, width, height, width/2, height/2)
}

// renderElement 渲染元素
// renderElement 渲染元素
func (s *SVGAdapter) renderElement(el SVGElement) string {
    scale := float64(displayScale)  // ← 改为 float64

    switch el.Type {
    case "image":
        href, _ := el.Data["href"].(string)
        if href == "" {
            return ""
        }

        centerX := (el.X + el.Width/2) * scale
        centerY := (el.Y + el.Height/2) * scale

        return fmt.Sprintf(`<image href="%s" x="%.2f" y="%.2f" width="%.2f" height="%.2f" preserveAspectRatio="none" transform="rotate(%.2f, %.2f, %.2f)" />`,
            escapeXML(href),
            el.X*scale,
            el.Y*scale,
            el.Width*scale,
            el.Height*scale,
            el.Rotation,
            centerX,
            centerY)

    case "text":
        text, _ := el.Data["text"].(string)
        size, _ := el.Data["size"].(float64)

        return fmt.Sprintf(`<text x="%.2f" y="%.2f" font-size="%.2f" font-family="Arial">%s</text>`,
            el.X*scale,
            el.Y*scale,
            size*scale,
            escapeXML(text))

    case "line":
        x1, _ := el.Data["x1"].(float64)
        y1, _ := el.Data["y1"].(float64)
        x2, _ := el.Data["x2"].(float64)
        y2, _ := el.Data["y2"].(float64)
        width, _ := el.Data["width"].(float64)
        color, _ := el.Data["color"].(string)
        dash, _ := el.Data["dash"].([]float64)

        dashAttr := ""
        if len(dash) > 0 {
            dashStrs := []string{}
            for _, d := range dash {
                dashStrs = append(dashStrs, fmt.Sprintf("%.2f", d*scale))
            }
            dashAttr = fmt.Sprintf(` stroke-dasharray="%s"`, strings.Join(dashStrs, ","))
        }

        return fmt.Sprintf(`<line x1="%.2f" y1="%.2f" x2="%.2f" y2="%.2f" stroke="%s" stroke-width="%.2f"%s />`,
            x1*scale, y1*scale, x2*scale, y2*scale, color, width*scale, dashAttr)

    case "rect":
        color, _ := el.Data["color"].(RGBColor)

        return fmt.Sprintf(`<rect x="%.2f" y="%.2f" width="%.2f" height="%.2f" fill="rgb(%d,%d,%d)" />`,
            el.X*scale,
            el.Y*scale,
            el.Width*scale,
            el.Height*scale,
            color.R, color.G, color.B)

    default:
        return ""
    }
}


// escapeXML 转义 XML 特殊字符
func escapeXML(s string) string {
    s = strings.ReplaceAll(s, "&", "&amp;")
    s = strings.ReplaceAll(s, "<", "<")
    s = strings.ReplaceAll(s, ">", ">")
    s = strings.ReplaceAll(s, "\"", "&quot;")
    s = strings.ReplaceAll(s, "'", "&apos;")
    return s
}

// ToBase64 将 SVG 转换为 base64
func ToBase64SVG(svg string) string {
    encoded := base64.StdEncoding.EncodeToString([]byte(svg))
    return fmt.Sprintf("data:image/svg+xml;base64,%s", encoded)
}
